File: kudos-ao3-reader/Services/WorkSearchIndex.swift
import Foundation
import OSLog
import SwiftData

/// Builds and queries `SavedWork.searchText` — a derived, normalized (lowercased,
/// diacritic-folded) concatenation of a work's searchable fields, computed once per
/// content change instead of per keystroke over the whole library.
///
/// The index is never source of truth: it's excluded from `.kudosbackup` exports,
/// and any record can be rebuilt from the real fields (`rebuildIfNeeded` sweeps
/// records whose `searchIndexVersion` doesn't match `currentVersion` at launch,
/// which also covers records created by pre-index builds or backup restores that
/// somehow missed their explicit reindex). Building the index never triggers a
/// network request — it reads only what's already stored locally.
@MainActor
enum WorkSearchIndex {
    /// Bump when `indexText(for:)`'s composition changes so existing records
    /// rebuild on next launch. 0 is reserved for "never indexed".
    /// v2: added the series title and the user's own tags (`SavedWork.tags`).
    static let currentVersion = 2

    /// Summary text beyond this contributes noise, memory, and store size, not
    /// recall — AO3 blurbs are typically well under it.
    private static let summaryLimit = 600

    /// Yield to the run loop every N records so a large stale library doesn't
    /// freeze the UI for the whole launch-time rebuild.
    private static let yieldInterval = 200

    // MARK: Normalization

    /// One shared normalization for both indexed text and queries: lowercased and
    /// diacritic-insensitive ("Héroïne" matches "heroine"), with a fixed locale so
    /// text indexed under one device locale matches queries typed under another.
    nonisolated static func normalize(_ text: String) -> String {
        text.folding(
            options: [.caseInsensitive, .diacriticInsensitive],
            locale: Locale(identifier: "en_US_POSIX")
        )
    }

    /// A query split into normalized terms. Matching is AND-across-terms, so
    /// "granger fluff" finds a work whose title has one word and tags the other —
    /// a single-term query behaves exactly like the old substring match.
    nonisolated static func terms(from query: String) -> [String] {
        normalize(query)
            .split(whereSeparator: \.isWhitespace)
            .map(String.init)
    }

    // MARK: Indexing

    /// The normalized searchable text for a work's current fields.
    static func indexText(for work: SavedWork) -> String {
        var parts: [String] = [work.title, work.author]
        // Series title, so a work is findable by the series it belongs to.
        parts.append(work.seriesTitle)
        // The user's own organizational tags. Every mutation site must reindex:
        // WorkDetailView's add/remove, and KudosBackup's restore (which links
        // archived user tags and must therefore reindex *after* linking them).
        parts.append(contentsOf: work.tags.map(\.name))
        parts.append(contentsOf: work.workFandoms)
        parts.append(contentsOf: work.workRelationships)
        parts.append(contentsOf: work.workCharacters)
        parts.append(contentsOf: work.workFreeforms)
        parts.append(contentsOf: work.workWarnings)
        parts.append(contentsOf: work.workCategories)
        // Flat tags cover works not yet enriched with categorized AO3 tags (e.g.
        // fresh EPUB imports) without duplicating every categorized entry above —
        // duplicates are harmless for matching but waste store space.
        let categorized = Set(
            (work.workFandoms + work.workRelationships + work.workCharacters
                + work.workFreeforms + work.workWarnings + work.workCategories)
                .map(normalize)
        )
        parts.append(contentsOf: work.workTags.filter { !categorized.contains(normalize($0)) })
        parts.append(work.rating)
        parts.append(work.language)
        parts.append(work.isComplete ? "complete" : "wip in progress")
        if !work.summary.isEmpty {
            parts.append(String(work.summary.strippingHTML().prefix(summaryLimit)))
        }
        return normalize(parts.filter { !$0.isEmpty }.joined(separator: "\n"))
    }

    /// Recomputes a work's derived search fields. Deliberately does **not** call
    /// `markModified()` — the index is derived state, and bumping the sync
    /// timestamp for it would make every reindex look like a user edit to the
    /// backup/folder-sync merge rules. Callers save the context themselves.
    static func reindex(_ work: SavedWork) {
        work.searchText = indexText(for: work)
        work.searchIndexVersion = currentVersion
    }

    /// Whether a work matches every normalized term (see `terms(from:)`). An empty
    /// term list matches everything, mirroring "no query".
    static func matches(_ work: SavedWork, terms: [String]) -> Bool {
        terms.allSatisfy { work.searchText.contains($0) }
    }

    // MARK: Rebuild

    /// Reindexes every record whose stamp doesn't match `currentVersion` — new
    /// installs' pre-index libraries, records restored from backups, and records
    /// indexed under an older schema. Cheap no-op when everything is current.
    @discardableResult
    static func rebuildIfNeeded(in context: ModelContext) async -> Int {
        let version = currentVersion
        let stale = (try? context.fetch(
            FetchDescriptor<SavedWork>(predicate: #Predicate { $0.searchIndexVersion != version })
        )) ?? []
        guard !stale.isEmpty else { return 0 }
        var processed = 0
        for work in stale {
            // The yields below let other main-actor work interleave — including a
            // user deleting one of the fetched works. Touching an invalidated model
            // asserts, so re-check liveness (same guard as the migration service).
            guard work.modelContext != nil else { continue }
            reindex(work)
            processed += 1
            if processed.isMultiple(of: yieldInterval) { await Task.yield() }
        }
        try? context.save()
        Log.library.info("Search index rebuilt for \(stale.count, privacy: .public) work(s)")
        return stale.count
    }
}

File: android/app/src/main/java/io/github/cidy02/kudos/works/WorkSearchIndex.kt
package io.github.cidy02.kudos.works

import io.github.cidy02.kudos.core.model.SavedWork
import java.text.Normalizer
import java.util.Locale
import kotlinx.coroutines.delay

/**
 * Normalization and search index for [SavedWork].
 * Port of Apple `WorkSearchIndex` (Services/WorkSearchIndex.swift).
 *
 * The index is derived state: excluded from backup, rebuilt on schema bump via
 * [rebuildIfNeeded], and recomputed on every upsert that changes searchable fields.
 */
object WorkSearchIndex {
    /**
     * Bump when [indexText] composition changes so existing records rebuild.
     * 0 is reserved for "never indexed".
     * v2: series title + user tags (parity with iOS v2).
     */
    const val CURRENT_VERSION = 2

    /** Summary text beyond this limit contributes noise, not recall. */
    private const val SUMMARY_LIMIT = 600

    /** Launch-rebuild pacing (iOS sliceBudget / sliceBreather / saveInterval). */
    private const val SLICE_BUDGET_MS = 5L
    private const val SLICE_BREATHER_MS = 2L
    private const val SAVE_INTERVAL = 250

    /**
     * Normalizes text for matching: lowercased, whitespace-trimmed, and diacritic-insensitive.
     */
    fun normalize(text: String): String {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return ""
        val normalized = Normalizer.normalize(trimmed, Normalizer.Form.NFD)
        return normalized.replace(Regex("\\p{InCombiningDiacriticalMarks}+"), "")
            .lowercase(Locale.ROOT)
    }

    /** Splits a query into normalized terms for AND matching across fields. */
    fun terms(query: String): List<String> {
        return normalize(query)
            .split(Regex("\\s+"))
            .filter { it.isNotBlank() }
    }

    /**
     * Generates the complete searchable text block for a work.
     * [userTags] are the user's organizational tags (iOS `SavedWork.tags`).
     */
    fun indexText(work: SavedWork, userTags: List<String> = emptyList()): String {
        val parts = mutableListOf<String>()
        if (work.title.isNotBlank()) parts.add(work.title)
        if (work.author.isNotBlank()) parts.add(work.author)
        if (work.seriesTitle.isNotBlank()) parts.add(work.seriesTitle)
        parts.addAll(userTags.filter { it.isNotBlank() })
        parts.addAll(work.workFandoms)
        parts.addAll(work.workRelationships)
        parts.addAll(work.workCharacters)
        parts.addAll(work.workFreeforms)
        parts.addAll(work.workWarnings)
        parts.addAll(work.workCategories)

        val categorized = (work.workFandoms + work.workRelationships + work.workCharacters +
            work.workFreeforms + work.workWarnings + work.workCategories)
            .map { normalize(it) }
            .toSet()

        parts.addAll(work.workTags.filter { normalize(it) !in categorized })
        if (work.rating.isNotBlank()) parts.add(work.rating)
        if (work.language.isNotBlank()) parts.add(work.language)
        parts.add(if (work.isComplete) "complete" else "wip in progress")
        if (work.summary.isNotBlank()) {
            val stripped = work.summary.replace(Regex("<[^>]*>"), "")
            parts.add(stripped.take(SUMMARY_LIMIT))
        }
        return normalize(parts.joinToString("\n"))
    }

    /**
     * Returns [work] with [SavedWork.searchText] / [SavedWork.searchIndexVersion] filled.
     * Does **not** bump `lastModifiedAt` — the index is derived state.
     */
    fun reindex(work: SavedWork, userTags: List<String> = emptyList()): SavedWork {
        return work.copy(
            searchText = indexText(work, userTags),
            searchIndexVersion = CURRENT_VERSION
        )
    }

    /**
     * True if [work] matches every term (AND).
     *
     * Prefer persisted [SavedWork.searchText] when current. [userTags] and
     * [extraTerms] (e.g. collection names) are always appended so library
     * search still finds membership-only tokens.
     */
    fun matches(
        work: SavedWork,
        terms: List<String>,
        userTags: List<String> = emptyList(),
        extraTerms: List<String> = emptyList()
    ): Boolean {
        if (terms.isEmpty()) return true
        val base = if (work.searchIndexVersion == CURRENT_VERSION && work.searchText.isNotEmpty()) {
            work.searchText
        } else {
            indexText(work, userTags)
        }
        val haystack = if (userTags.isEmpty() && extraTerms.isEmpty()) {
            base
        } else {
            // Re-include userTags when recomputing wasn't done; extras always needed.
            val extras = buildList {
                if (work.searchIndexVersion != CURRENT_VERSION || work.searchText.isEmpty()) {
                    // already in base via indexText
                } else {
                    addAll(userTags)
                }
                addAll(extraTerms)
            }
            if (extras.isEmpty()) base else normalize(base + "\n" + extras.joinToString("\n"))
        }
        return terms.all { haystack.contains(it) }
    }

    /**
     * Reindexes every work whose stamp doesn't match [CURRENT_VERSION].
     * Paced so large libraries don't stall. Returns count rebuilt.
     */
    suspend fun rebuildIfNeeded(
        loadStale: suspend (currentVersion: Int) -> List<SavedWork>,
        userTagsFor: suspend (workId: String) -> List<String>,
        save: suspend (List<SavedWork>) -> Unit
    ): Int {
        val stale = loadStale(CURRENT_VERSION)
        if (stale.isEmpty()) return 0
        val pending = mutableListOf<SavedWork>()
        var sinceSave = 0
        var sliceStart = System.nanoTime()
        for (work in stale) {
            val tags = userTagsFor(work.id)
            pending.add(reindex(work, tags))
            sinceSave++
            if (sinceSave >= SAVE_INTERVAL) {
                save(pending.toList())
                pending.clear()
                sinceSave = 0
            }
            val elapsedMs = (System.nanoTime() - sliceStart) / 1_000_000L
            if (elapsedMs > SLICE_BUDGET_MS) {
                delay(SLICE_BREATHER_MS)
                sliceStart = System.nanoTime()
            }
        }
        if (pending.isNotEmpty()) {
            save(pending)
        }
        return stale.size
    }
}
